### ISSUE TITLE
Running together with Timestamper prevents AnsiColor 0.7.0 from working correctly, 0.6.3 ok

### ISSUE THREAD

**Issue by alexharv074** [jenkinsci/ansicolor-plugin#193]
### Test environment

- Jenkins Version:  2.176.3
- AnsiColor Version: 0.7.0

### How I installed

Showing it is installed:
<img width="1176" alt="Screen Shot 2020-06-12 at 3 24 52 am" src="https://user-images.githubusercontent.com/2305330/84420222-1277d280-ac5d-11ea-81af-0e2803b52eda.png">

I also restarted Jenkins (several times).

I have a build step that looks like this:

```groovy
         wrap([$class: 'AnsiColorBuildWrapper', 'colorMapName': 'XTerm']) {
            sh "[ ${params.RUN_E2E} = true ] && (cd /src; TEMPLATE=${params.TEMPLATE} make e2e) || true"
          }
```

I also tried:

```groovy
   options {
     ansiColor('xterm')
  }
```

### Expected behavior

I expected coloured output from Sceptre and shUnit2.

### Actual behavior

It's not coloured:
<img width="378" alt="Screen Shot 2020-06-12 at 3 29 38 am" src="https://user-images.githubusercontent.com/2305330/84420174-01c75c80-ac5d-11ea-9268-7249c63927e4.png">

**Comment by alexharv074** [comment]
For any tips for anything else I could try I would be most grateful!

**Comment by tszmytka** [comment]
Thanks for the report and providing detailed information. Unfortunately it is not immediately obvious what might be wrong. Could you then please:
1. Set up a job with the following pipeline. Run it and post your console output.
```
ansiColor('xterm') {
    echo '\033[32mHello World\033[0m'
}
```

2. Run your test locally, capture and post the output. I'm thinking something like:
```
make e2e > output.txt
```

**Comment by alexharv074** [comment]
@tszmytka yeah that works 
<img width="522" alt="Screen Shot 2020-06-13 at 12 02 26 am" src="https://user-images.githubusercontent.com/2305330/84511128-6a6c1300-ad09-11ea-8f96-ab83f8250b11.png">

I also attached here the output file:
[output.txt](https://github.com/jenkinsci/ansicolor-plugin/files/4771247/output.txt)

Also my Jenkinsfile:
[Jenkinsfile.groovy.txt](https://github.com/jenkinsci/ansicolor-plugin/files/4771265/Jenkinsfile.groovy.txt)

**Comment by alexharv074** [comment]
@tszmytka Is it because I am using the Jenkinsfile Groovy plugin https://plugins.jenkins.io/workflow-cps/ ?

**Comment by tszmytka** [comment]
Basic `Hello World` example works which means the plugin is loaded and running. 

>I also attached here the output file:
>output.txt

The output doesn't seem out of the ordinary.

>Also my Jenkinsfile:
>Jenkinsfile.groovy.txt

What plugin does the `container('awscli')` step come from? This is not a standard pipeline way of running docker images.

>Is it because I am using the Jenkinsfile Groovy plugin https://plugins.jenkins.io/workflow-cps/ ?

I doubt it. It is used by probably all people utilizing pipelines so if there was an incompatibility with `ansicolor` we would have gotten a ton of bugtickets.

**Comment by joshdcox** [comment]
FWIW, AnsiColor 0.7.0 isn't working for us either. We reverted to 0.6.3, which still works works fine.

**Comment by alexharv074** [comment]
Ok, many thanks to @joshdcox for that info. I have also confirmed that by downgrading to 0.6.3, the colouring is working fine for me.

**Comment by fishi0x01** [comment]
Same for me (downgrade to `0.6.3` works). 

On `0.7.0` I have noticed that the log coloring works properly while the build is still running. When looking at logs of builds which have already finished the log coloring vanished.

**Comment by tszmytka** [comment]
I am still unable to reproduce the issue. I suspect (could be wrong) this has everything to do with jobs running on specific agents, although my local setup (with multiple agents) runs all the test cases fine.
I've prepared a simple pipeline
```
pipeline {
  agent {
    label "<your-agent>"
  }

  options {
    ansiColor('xterm')
  }

  stages {
    stage('Run') {
      steps {
          echo '\033[32mcolored output\033[0m'
          sh "echo '\\033[92;22mlightgreen and combined reset sgr \\033[0m'"
      }
    }
  }
}
```

@alexharv074 Would you be able to run this against your `sceptre-builder` agent and post results? Also, what kind of agent is run on that host? I was testing the plugin using [the official inbound-agent docker image](https://hub.docker.com/r/jenkins/inbound-agent/) and still could not reproduce so I'm assuming you're using something else. Can you provide the name of the image?

@joshdcox and @fishi0x01 Thanks for your additional information. Could you please confirm that you are also running your jobs against a remote agent? If so could you also please run my test pipeline and post results?

**Comment by UzK98fYoE** [comment]
I'm hitting the same issue too on a CentOS 7.7 system when running an ansible container running python 3.8. It was working fine on my CentOS 7.6 system until I migrated to the newer one. tszmytka's code works however. Troubleshooting now...

**Comment by tszmytka** [comment]
Another idea: Could you please temporarily restart your **master** with the following JVM option:
```
-Dorg.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep.USE_WATCHING=false
```
run your job again and report if this changes anything?

**Comment by reinholdfuereder** [comment]
(I am very happy with AnsiColor 0.7.0: colours are shown as expected in classic and Blue Ocean UI)

**Comment by faithlinpaul** [comment]
Doesn't work for me with version`0.7.0` and `Jenkins 2.235.1`

**Comment by ant100** [comment]
It's not working for me either on version 0.7.0 and Jenkins 2.235.1

**Comment by eightnoneone** [comment]
Maybe small detail but I'm seeing that with truncated console output, the colors are not displayed. If I view full log, it displays correctly.

**Comment by chder** [comment]
Using the test case in https://github.com/jenkinsci/ansicolor-plugin/issues/193#issuecomment-650812651 I see colors render on Jenkins 2.235.1.
But, if I add to the options block `timestamps()` it breaks.

Here's the raw log output showing a part that should be colorized run through xxd to show where the control sequences are:
```
$ grep --color=none lightgreen log | grep -v 'echo'
[2020-07-06T16:51:19.843Z] lightgreen and combined reset sgr 
$ grep --color=none lightgreen log | grep -v 'echo' | xxd
0000000: 5b32 3032 302d 3037 2d30 3654 3136 3a35  [2020-07-06T16:5
0000010: 313a 3139 2e38 3433 5a5d 201b 5b39 323b  1:19.843Z] .[92;
0000020: 3232 6d6c 6967 6874 6772 6565 6e20 616e  22mlightgreen an
0000030: 6420 636f 6d62 696e 6564 2072 6573 6574  d combined reset
0000040: 2073 6772 201b 5b30 6d0a                  sgr .[0m.
$
```
Here's a screenshot:
![image](https://user-images.githubusercontent.com/5008161/86619573-7b164e80-bf80-11ea-93e2-a92ff6db6978.png)

But even after a refresh in Jenkins, not rendering color:
![image](https://user-images.githubusercontent.com/5008161/86619668-a600a280-bf80-11ea-8760-24c610f83db5.png)

Pipeline ran:
```
// Test pipeline from https://github.com/jenkinsci/ansicolor-plugin/issues/193#issuecomment-650812651
pipeline {
  agent any

  options {
    timestamps()
    ansiColor('xterm')
  }

  stages {
    stage('Run') {
      steps {
          echo '\033[32mcolored output\033[0m'
          sh "echo -e '\\033[92;22mlightgreen and combined reset sgr \\033[0m'"
      }
    }
  }
}

```

**Comment by dayer4b** [comment]
I just wanted to join the chorus of voices on this issue.  0.7.0 does not work for us.  I rolled back and it works fine.  Here's the test `Jenkinsfile`:

```groovy
node {
  //timestamps
  timestamps{
    // This displays colors using the 'xterm' ansi color map.
    ansiColor('xterm') {
    stage ("Look for the colors!"){
      echo "\u001B[31mI'm Red\u001B[0m Now not"
      sleep 10
      echo "\u001B[31mI'm Red\u001B[0m Now not"
    }
    }
  }
}
```

With 0.6.2 this displays red for the "I'm Red" part, with 0.7.0 it prints like this:
```
09:28:22  [31mI'm Red[0m Now not
```

**Comment by tszmytka** [comment]
@chder You hit the nail right on the head. Thank you a thousand times.
> But, if I add to the options block timestamps() it breaks.

The `timestamper-plugin` is the missing piece of the puzzle I've been searching for. Adding it to the pipeline prevents any escape sequences from being correctly interpreted. I am able to reproduce it in a consistent manner and reverting back to `ansicolor vansicolor-0.6.3` brings the functionality back.
In fact looking **only now** at the OP's "output.txt" from https://github.com/jenkinsci/ansicolor-plugin/issues/193#issuecomment-643291231 also shows dates and times so he most likely is also using `timestamper`.

Quick tests on my dev setup show that while the following
```
options {
  timestamps()
  ansiColor('xterm')
}
```
breaks the escape codes, 
this:
```
options {
  ansiColor('xterm')
  timestamps()
}
```
does not (in `v0.7.0`). 
Suggest to try this as a quick work around. In the mean time I will try to find a permanent solution.

**Comment by maemile** [comment]
Hello,

Same issue here...
The workaround fix only in fully expanded in my case and not in truncated.

**Comment by tszmytka** [comment]
@maemile Let's stick to the main issue here.
> The workaround fix only in fully expanded in my case and not in truncated.

Problem with rendering escape sequences on partial output is described in #185 .

**Comment by tszmytka** [comment]
The problem in this ticket is caused by the `timestamper` plugin decorating the logger in such a way that **only** lines which end with a `LF` character get written to the log file. Since `ansicolor v0.7.0` we started annotating log lines with meta data in order to better work in situations where say, you want to use multiple color maps within one pipeline or you only want part of your output colored.

The actual fix will also need to be done in `timestamper` but till that happens `ansicolor v0.7.1` will try to detect this situation and manually force the logger to actually print contents to the file. This will have a side effect of having an additional, empty line in the output but at least main functionality will get back to normal.

**Comment by tszmytka** [comment]
The newly released `ansicolor v0.7.1` contains a method to work around `timestamper` swallowing up meta-data info (and thus disabling escape character recognition).

All is working fine on my dev setup but - as this was pretty hard to analyze in the wild - could any one of the reporters please review this on your system and confirm it is working for you too?
Please make sure you are using both plugins:
- ansicolor v0.7.1
- timestamper (any version)

**Comment by patricks** [comment]
I have updated to `v0.7.1` but I didn't see any difference.

`[08:32:19]: [35m[33m▸[0m [39;1mBuild[0m Succeeded[0m`

**Comment by tszmytka** [comment]
How do I reproduce this?
Care to provide a sample pipeline showing how you use both plugins?
Also, which version of Jenkins and timestamper are you using?

**Comment by tszmytka** [comment]
Going through `timestamper`s code I found it can be enabled also without explicitly setting it in the pipeline (through a global configuration).

@patricks Is this how you're using it?

**Comment by patricks** [comment]
I just exec a shell command to run a ruby script in my pipeline.

```
ansiColor("xterm") {
    sh "bundle exec fastlane test"
 }
```

**Comment by mscholze** [comment]
Colors still not rendered, Jenkins 2.235.2, AnsiColor 0.7.1, Timestamper 1.11.3.

**Comment by marslo** [comment]
my case:
Jenkins: `2.2.245`
timestamps: `1.11.3`
ansiColor: `0.7.1`

Colors cannot show up by:
```
ansiColor('xterm') { timestamps {
...
}}
```

but works in :
```
timestamps { ansiColor('xterm') {
...
}}
```

**Comment by mscholze** [comment]
@marslo It looks like scripted pipeline. But good to know.

**Comment by dblock** [comment]
I would like to encourage someone from this thread to step up to being much more active in debugging, testing and fixing new/existing bugs. Write more unit tests. Try HEAD right now after https://github.com/jenkinsci/ansicolor-plugin/pull/198 has been merged. Right now @tszmytka is doing solo work here with barely another pair of eyes on the code (I am not actively using Jenkins anymore, and am not familiar with recent versions and internals). This project needs another highly engaged programmer/co-maintainer to avoid having so many regressions.

**Comment by dblock** [comment]
I also want to say a *huge* thank you to @tszmytka for carrying this forward!

**Comment by tszmytka** [comment]
I just released `v0.7.2` containing 2 important fixes that will hopefully help to sort out this issue.
I've tested the interaction between `ansicolor` and `timestamper` in all ways known to mankind and could see no problems. Here's the list of configurations I used:
1. By using pipeline-wide wrappers
```
options {
    timestamps()
    ansiColor('xterm')
}
```

and

```
options {
    ansiColor('xterm')
    timestamps()
}
```

2. By using ad-hoc steps:

```
ansiColor('xterm') {
    timestamps {
	
	}
}
```
and
```
timestamps {
    ansiColor('xterm')  {
	
	}
}
```
3. By enabling both plugins in Jenkins' global options.


A small suggestion to anyone suspecting they are suffering from this issue (as till now I can identify people reporting 3 separate issues in this thread - 2 of which fixed): Try **disabling** `timestamper` **entirely** (that means no `timestamps` in pipeline and no global `Enabled for all Pipeline builds` setting) and run your pipeline again. If the output is colored then you are - please comment on this thread providing **details** of how to reproduce what you are seeing. 
But if your output is still not colored or you are not using `timestamper` then your issue is something else - please open up a separate ticket.

**Comment by patricks** [comment]
It looks like the latest `timestamper` and `ansicolor` updates fixed my problem. 👏

**Comment by mscholze** [comment]
I can confirm the fix 👍 

tested declarative pipeline on:
- jenkins 2.235.3
- ansicolor 0.7.2
- timestamper 1.11.5

**Comment by asantos-fuze** [comment]
Whatever fix was made it is  now working

- `jenkins 2.235.2`
- `ansicolor 0.7.2`
- `timestamper 1.11.5`

**Comment by chder** [comment]
It's mostly fixed for me but there's still a situation where it's not so I made up a somewhat contrived pipeline script.

```groovy
// Test pipeline for https://github.com/jenkinsci/ansicolor-plugin/issues/193
pipeline {
  agent any

  options {
    timestamps()
    ansiColor('xterm')
  }

  stages {
    stage('Run') {
      steps {
          echo '\033[32mcolored output\033[0m'
sh label: 'colors', script: '''echo start
for c in {30..37}; 
do
  printf "Time for $c\\n"
  sleep 3
  echo -e "\\e[31mHello World\\e[0m"
  echo -e "\\e[${c}m Loop $c \\e[0m"
  printf '\\033[32mcolored output\\033[0m\\n'
  printf "\\033[${c}mcolored output\\033[0m\\n" 
  for i in {000..199}
  do 
    printf "\\033[${c}mI\\\'m filler $i taking up a bunch of log space ####################################################################################\\033[0m\\n" 
  done
  echo next
done'''
    
    
      }
    }
  }
}
```

The problem for me now is when Jenkins truncates the output when it gets too long. 
After starting that pipeline mash refresh on the console output view every ~3 seconds. It's looping through colors 30 to 37 and at about 33 (refresh when it's on green) refreshing loses the color. I suspect it's hitting a case were an escape sequence is getting split by the console getting truncated and whatever is doing the parsing decides not to try past the invalid input but just a guess.
___
![Peek 2020-08-05 10-40 ansiColor](https://user-images.githubusercontent.com/5008161/89433584-21807b80-d708-11ea-8bb5-525ec9caaca0.gif)
Example where I hit refresh on the green step (there colors stop working) and then after it finishes (color works again).

Versions:
```
Jenkins 2.235.1
ansicolor 0.7.2
Timestamper 1.11.3
```

**Comment by julian-alarcon** [comment]
https://github.com/jenkinsci/ansicolor-plugin/pull/196

**Comment by tszmytka** [comment]
@chder Thanks for a very nice pipeline. It has nothing to do with `Timestamper` so I moved it to a separate issue (#200).

Other than that I think it's clear to assume this issue has been fixed in `v0.7.2`. 
Thank you to everyone who invested their time and tested various combinations.

**Comment by jjathman** [comment]
We seem to still be having problems. Our pipelines use this pattern:

```
options {
    timestamps()
}

// extra detailed removed

ansiColor('xterm') {
   // do work here
}
```

Should this use case be working?
